### ISSUE TITLE
Improve the Loading Time for the Author Manager

### ISSUE THREAD

**Issue by twinkietoes-on** [LibriVox/librivox-catalog#24]
Some of the admins have reported that the Author Manager doesn't load for them - it fails before it gets to the page of unconfirmed authors with the search box - so they can't edit or add new authors. Is there something we can tweak to make this table load faster, or at least keep it from timing out and failing?

Priority: 1

**Comment by hamidsafdari** [comment]
I have implemented pagination for this page. #176 improves the loading time. It also adds search and ordering. Everything is reflected in the page URL and restored on page refresh.

This is the before:

[issue-24-before.webm](https://github.com/LibriVox/librivox-catalog/assets/5158006/b1e3002d-198e-4f51-9aae-8d024ded7b4b)

This is the after:

[issue-24-after.webm](https://github.com/LibriVox/librivox-catalog/assets/5158006/3de0243b-5092-42e9-908d-8b02e0dcaf82)

**Comment by twinkietoes-on** [comment]
From watching the video:
- So all the records are loaded into the browser, so one doesn't have to reload to get more records, correct? (Upon load, all authors are findable by the search, etc.?)
- The "Linked" option - it's interesting to have them visible, but I'm not sure it won't be confusing. The original designer had an aversion to deleting any records, so he built this in as a way to combine duplicate records. I'm not sure we need to be able to see them separated out and visible. At any rate, this would require some discussion as to whether or not we'd want that feature.

    On Tuesday, November 21, 2023 at 11:51:07 a.m. EST, Hamid Safdari ***@***.***> wrote:  
 
 


I have implemented pagination for this page. #176 improves the loading time. It also adds search and ordering. Everything is reflected in the page URL and restored on page refresh.

This is the before:

issue-24-before.webm

This is the after:

issue-24-after.webm

**Comment by hamidsafdari** [comment]
> - So all the records are loaded into the browser, so one doesn't have to reload to get more records, correct? (Upon load, all authors are findable by the search, etc.?)

No. That was causing the slow page load in the first place. I'm setting a "start" and "length" param in the URL that will be used in the code to know which records to load. On my laptop, using the local database, the original page is 2.58MBs and loads in 13.5 seconds. I don't know how many records are in production but the load time there is definitely much more than what I can do locally. The new paginated page is 8KBs and loads in 300ms. There are ~12000 records loaded into memory in the original page which is by no measure a good thing to do. The only options are to either use params in the URL to track where you are (even on reload) or to transfer data over AJAX. The first option seemed easier to do. There are old libraries used and there's no distinction b/w the frontend and backend code. This would be a lot easier if they were separated.

> - The "Linked" option - it's interesting to have them visible, but I'm not sure it won't be confusing. The original designer had an aversion to deleting any records, so he built this in as a way to combine duplicate records. I'm not sure we need to be able to see them separated out and visible. At any rate, this would require some discussion as to whether or not we'd want that feature.

That can be removed. So what is "not" linked is deleted and should not be displayed?

**Comment by hamidsafdari** [comment]
The linked dropdown is now removed and a default value of 0 is assumed.

**Comment by twinkietoes-on** [comment]
My programmer husband said something about the loading of the records into the browser isn't what takes up the time; it's the crappy Javascript handling of the records after that. (He looked at it a few years ago, but work gets in the way of doing much with the system nowadays.)
As long as our functionality remains essentially the same - being able to find duplicates and such - then it may be good with the change. It's hard to predict what the side effects will be - unintended consequences and all that. 

We're visiting family for Thanksgiving, so I don't have much time to dig into this or play with anything in a sandbox environment. Your videos looked pretty good, but it doesn't beat getting one's own hands on it to try it for oneself. :)
I guess my ultimate question is, if this should happen to have negative unforeseen effects, would it be easy to reverse the change? If so,I'm happy to see it implemented.

    On Tuesday, November 21, 2023 at 08:34:08 p.m. EST, Hamid Safdari ***@***.***> wrote:  
 
 


   
   - So all the records are loaded into the browser, so one doesn't have to reload to get more records, correct? (Upon load, all authors are findable by the search, etc.?)


No. That was causing the slow page load in the first place. I'm setting a "start" and "length" param in the URL that will be used in the code to know which records to load. On my laptop, using the local database, the original page is 2.58MBs and loads in 13.5 seconds. I don't know how many records are in production but the load time there is definitely much more than what I can do locally. The new paginated page is 8KBs and loads in 300ms. There are ~12000 records loaded into memory in the original page which is by no measure a good thing to do. The only options are to either use params in the URL to track where you are (even on reload) or to transfer data over AJAX. The first option seemed easier to do. There are old libraries used and there's no distinction b/w the frontend and backend code. This would be a lot easier if they were separated.

   
   - The "Linked" option - it's interesting to have them visible, but I'm not sure it won't be confusing. The original designer had an aversion to deleting any records, so he built this in as a way to combine duplicate records. I'm not sure we need to be able to see them separated out and visible. At any rate, this would require some discussion as to whether or not we'd want that feature.


That can be removed. So what is "not" linked is deleted and should not be displayed?

**Comment by hamidsafdari** [comment]
To reverse it, one could just get the commit hash and do `git revert [hash]`. That would generate a new commit that reverses whatever was done.
